### PR TITLE
cmd/abigen, accounts/abi/bind: options for customising generated code

### PR DESCRIPTION
**Pull request by aschlosberg**
The code generated by `bind.Bind()` and `abigen` can currently only be modified via the `ast` package (Go only), which is difficult, or through string manipulation, which is brittle. This PR exposes functions to override the default code templates through `BindOption`s. The `WithTemplate()` option is also exposed via `abigen --custom-template`.

The variadic addition to the `bind.Bind()` argument list leaves the function signature unchanged, thus keeping backwards compatibility. The `BindOption` deliberately can't be implemented outside of the package due to an unexported field in the interface signature; this is because `Bind()`, in its current form, needs to be aware of all possible options.

Both the Go and Java template-source constants are now exported to allow for modification that only requires extra generated code and not a full rewrite.

### CODE REVIEW

**Comment by holiman** [comment]
> Both the Go and Java template-source constants are now exported to allow for modification that only requires extra generated code and not a full rewrite.

I don't understand this. In what case would you have any use for them exported? You can only ever use them to build a new template by embedding them and adding some header or footer.. ?

File: accounts/abi/bind/bind.go [hunk 1]
```diff
@@ -47,7 +47,7 @@ const (
 // to be used as is in client code, but rather as an intermediate struct which
 // enforces compile time type safety and naming convention opposed to having to
 // manually maintain hard coded strings that break on runtime.
-func Bind(types []string, abis []string, bytecodes []string, fsigs []map[string]string, pkg string, lang Lang, libs map[string]string, aliases map[string]string) (string, error) {
+func Bind(types []string, abis []string, bytecodes []string, fsigs []map[string]string, pkg string, lang Lang, libs map[string]string, aliases map[string]string, opts ...BindOption) (string, error) {
```
Comment (holiman) [inline accounts/abi/bind/bind.go:50]: Thinking aloud here, without having reviewed/understood the entire method, but.. 
This is an api-breaking change, so would it make sense to instead add a different method
`BindWithOpts( x, y, z,  .. , opts ... BindOption) (stnrg, error)` instead? 

And possible bundle as much as possible of the _existing_ input arguments as `BindOption`s.

Comment (aschlosberg) [reply accounts/abi/bind/bind.go:50]: It's actually non-breaking by design. The variadic `...` allows the function to be called with zero or more individual `BindOption`s so all existing code remains unaffected—I think there's at least one call to `Bind()` in the Java tests that remains unchanged.

This is a common pattern in, for example, Google's gRPC https://pkg.go.dev/google.golang.org/grpc

**Comment by aschlosberg** [comment]
> > Both the Go and Java template-source constants are now exported to allow for modification that only requires extra generated code and not a full rewrite.
> 
> I don't understand this. In what case would you have any use for them exported? You can only ever use them to build a new template by embedding them and adding some header or footer.. ?

This is the one part of the PR I felt a bit unsure about. In my particular use case I did indeed want to add a footer, @holiman, but I certainly see why you're questioning it.

**Comment by holiman** [comment]
This PR is growing stale, and I do not think this will be merged -- it adds complexity, and the benefits do not seem to outweigh the overhead. 
Thanks for making the effort, and apologies, but I'm closing this.

**Pull request closed without merging**